src/PreserveModuleNamePlugin.ts

```diff
@@ -72,90 +164,320 @@ function getPreservedModules(modules: Webpack.Module[]) {
   return new Set(
     modules.filter(m => {
       // Some modules might have [preserveModuleName] already set, see ConventionDependenciesPlugin.
-      let value = m[preserveModuleName];      
-      for (let r of m.reasons) {
-        if (!r.dependency || !r.dependency[preserveModuleName]) continue;
-        value = true;
-        let req = removeLoaders((r.dependency as ModuleDependency).request);
-        // We try to find an absolute string and set that as the module [preserveModuleName], as it's the best id.
-        if (req && !req.startsWith(".")) {
-          m[preserveModuleName] = req;
-          return true;
-        }
+      if (m[preserveModuleName]) {
+        return true;
       }
-      return !!value;
+
+      // Preserve the module if its dependencies are also preserved
+      const reasons = (m.reasons && Array.isArray(m.reasons)) ? m.reasons : [];
+      return reasons.some((reason) => Boolean(reason.dependency && reason.dependency[preserveModuleName]));
```

Comment:
Nit: no need for parenthesis when there's a single argument. No need to convert to bool explicitely as it's done by `some`.